analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs

```diff
@@ -31,8 +33,39 @@ public class LocksReleasedAllPaths : LocksReleasedAllPathsBase
 
         protected override DiagnosticDescriptor Rule => S2222;
 
-        // ToDo: Implement early bail-out if there's no interesting descendant node in context.Node to avoid useless SE runs https://github.com/SonarSource/sonar-dotnet/issues/5375
-        public override bool ShouldExecute() =>
-            NodeContext.Node.DescendantNodes().OfType<IdentifierNameSyntax>().Any(x => ShouldExecuteFor(x.Identifier));
+        protected override void Visit(SyntaxNode node, LockAcquireReleaseCollector collector)
+        {
+            var walker = new LockAcquireReleaseWalker(collector);
+            walker.SafeVisit(NodeContext.Node);
+        }
+
+        private sealed class LockAcquireReleaseWalker : CSharpSyntaxWalker
+        {
+            private readonly LockAcquireReleaseCollector collector;
+
+            public LockAcquireReleaseWalker(LockAcquireReleaseCollector collector)
+            {
+                this.collector = collector;
+            }
+
+            public override void Visit(SyntaxNode node)
+            {
+                if (collector.LockAcquiredAndReleased
+                    // Lambda expressions and local functions are analyzed separately.
+                    || node is LambdaExpressionSyntax or ParenthesizedLambdaExpressionSyntax
```

Comment:
We need to cover all cases from here:
https://github.com/SonarSource/sonar-dotnet/blob/88af512b7bba6f3095c1d42f4b39d2b422d2799d/analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs#L135-L136

```suggestion
                    || node is LambdaExpressionSyntax or ParenthesizedLambdaExpressionSyntax or AnonymousMethodExpressionSyntax
```
Test case:
```
                System.VoidDelegate anonymousMethod = delegate     
                {
                    // Something
                };
```

Reply:
Good point. We can cover all of them using the base class `AnonymousFunctionExpressionSyntax`.